aten/src/ATen/native/cudnn/ConvShared.cpp

```diff
@@ -0,0 +1,500 @@
+  checkAllSameGPU(c, {input, weight});
+
+  auto layout = cudnn_conv_use_channels_last(*input, *weight) ?
+      at::MemoryFormat::ChannelsLast : at::MemoryFormat::Contiguous;
+  auto output_t = at::empty(
+                    conv_output_size(input->sizes(), weight->sizes(),
+                                     padding, stride, dilation),
+                    input->options(),
+                    layout);
+
+  if (output_t.numel() == 0) {
+    return output_t;
+  }
+
+  // Avoid ambiguity of "output" when this is being used as backwards
+  TensorArg output{ output_t, "result", 0 };
+  convolution_shape_check(c, input, weight, output, padding, stride, dilation, groups);
+
+  // See #4500
+  Tensor weight_contig = weight->contiguous(layout);
+  // Make sure that NC11 strides follow formula
+  weight_contig.resize_(weight_contig.sizes(), layout);
+  Tensor input_contig = input->contiguous(layout);
+  input_contig.resize_(input_contig.sizes(), layout);
+
+  raw_cudnn_convolution_forward_out(
+      *output, input_contig, weight_contig,
+      padding, stride, dilation, groups, benchmark, deterministic, allow_tf32);
+
+  return *output;
+}
+
+Tensor cudnn_convolution(
+    const Tensor& input_t, const Tensor& weight_t,
+    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation,
+    int64_t groups, bool benchmark, bool deterministic, bool allow_tf32)
+{
+  TensorArg input  { input_t,  "input",  1 },
+            weight { weight_t, "weight", 2 };
+  CheckedFrom c = "cudnn_convolution";
+  auto output_t = cudnn_convolution_forward(
+    c, input, weight, padding, stride, dilation, groups, benchmark, deterministic, allow_tf32);
+  return output_t;
+}
+
+// NB: output_padding not needed here, as there is no ambiguity to
+// resolve
+Tensor cudnn_convolution_transpose_backward_input(
+    const Tensor& grad_output_t, const Tensor& weight_t,
+    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation,
+    int64_t groups, bool benchmark, bool deterministic, bool allow_tf32)
+{
+  TensorArg grad_output { grad_output_t,  "grad_output", 1 },
+            weight      { weight_t, "weight", 2 };
+  return cudnn_convolution_forward(
+    "cudnn_convolution_transpose_backward_input",
+    grad_output, weight, padding, stride, dilation, groups, benchmark, deterministic, allow_tf32);
+}
+
+std::tuple<at::Tensor,at::Tensor> cudnn_convolution_transpose_backward(
+    const at::Tensor& input, const at::Tensor& grad_output_t, const at::Tensor& weight,
+    IntArrayRef padding, IntArrayRef output_padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups,
+    bool benchmark, bool deterministic, bool allow_tf32, std::array<bool,2> output_mask) {
+
+  Tensor grad_output = grad_output_t.contiguous(input.suggest_memory_format());
+
+  Tensor grad_input, grad_weight;
+  if (output_mask[0]) {
+    grad_input = at::cudnn_convolution_transpose_backward_input(grad_output, weight, padding, stride, dilation, groups, benchmark, deterministic, allow_tf32);
```

Comment:
Maybe you could call cudnn_convolution_forward directly here, and get rid of cudnn_convolution and cudnn_convolution_transpose_backward_input? Do they provide any value?

Reply:
I think the value is a better error message, i.e. correctly calling it `grad_output` instead of `input`